File: BOJ/29_유니온파인드/20040_사이클게임.cpp
#include <iostream>

#define MAXN 500000

using namespace std;

int par[MAXN+1];
int N,M,x,y;

int find(int x) {
    if(par[x] == x) return par[x];
    else return par[x] = find(par[x]);
}

void merge(int x, int y) {
    int p_x = find(x);
    int p_y = find(y);
    if(p_x == p_y) return ;
    par[p_x] = p_y;
}

int main() {
    
    scanf("%d %d",&N,&M);
    int ans = 0;
    for(int i=0; i<=N; i++) par[i] = i;
    for(int i=1; i<=M; i++) {
        scanf("%d %d",&x,&y);
        if(find(x) == find(y) && ans == 0) ans = i;
        merge(x,y);
    }
    printf("%d\n",ans);
    
    return 0;

}

File: BOJ/29_유니온파인드/20040_사이클게임.py
import sys

sys.setrecursionlimit(int(1e9))
read = sys.stdin.readline

def find(x) :
    if par[x] == x : return par[x]
    else :
        par[x] = find(par[x])
        return par[x]

def merge(x,y) :
    p_x = find(x)
    p_y = find(y)

    if p_x == p_y : return
    par[p_x] = p_y

N,M = map(int,read().split())
par = [i for i in range(N+1)]
end = 0
for i in range(1,M+1) :
    x,y = map(int,read().split())
    if find(x) == find(y) and end == 0 :
        end = i
    
    merge(x, y)

print(end)